### ISSUE TITLE
Dropout behavior similarly across GPUs

### ISSUE THREAD

**Issue by lqtrung1998** [huggingface/accelerate#114]
Hi, 

I'm referring to this question: https://discuss.pytorch.org/t/ddp-and-dropout-behave-the-same-across-gpus/125052

I'm not too sure if we can make Dropout in different GPUs independent  by setting seed after getting a batch and resynchronize RNG state across GPUs when processing the next batch.

https://github.com/huggingface/accelerate/blob/main/src/accelerate/data_loader.py#L286-L287

Correct me if I'm wrong but I see that rng_types is passed (without copy) every time new dataloader is put into prepare function. This may remove 'generator' from the global rng_types. Is this expected ? or a copy of rng_types should be used?

https://github.com/huggingface/accelerate/blob/main/src/accelerate/data_loader.py#L409-L410

Any help is appreciated! Thanks

**Comment by sgugger** [comment]
Oh that's a nasty bug. I think [this commit](https://github.com/huggingface/accelerate/commit/6608466d0ad96dd573a86ddaf2a91a21eccba5d1) should fix it, could you confirm?

**Comment by lqtrung1998** [comment]
Yes, I believe that would work.

Btw, does the current version support loading data in distributed manner ? Cause I see each process iterates over entire dataset and just returns a subset of it. If yes, can you show me how ... if not, then will it be supported soon ?

**Comment by sgugger** [comment]
What do you mean by "load data in a distributed manner"?
Each process sees a shard of the dataset, yes, that is one of the main things Accelerate implements.

**Comment by lqtrung1998** [comment]
What I mean is that if we use Distributed Sampler ([ref](https://pytorch.org/docs/stable/_modules/torch/utils/data/distributed.html#DistributedSampler)), each GPU only processes the indexes assigned to it (since the index returned by the sampler is `indices[self.rank:self.total_size:self.num_replicas]`. So each process will only call `__getitem__` and collate_fn on a subset of the dataset.

But from my understanding, in BatchSamplerShard, each GPU will call `__getitem__` and `collate_fn` on all entire dataset but just return the subset assigned to it ....

**Comment by sgugger** [comment]
No, it is, as the name indicates, a `BatchSampler`. It only computes indices (specifically the indices of the elements we want in each batch), the access to the Dataset is done later inside the `DataLoader`.

**Comment by lqtrung1998** [comment]
I'm clear now, thank you for the clarification.
